src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java

```diff
@@ -338,6 +358,47 @@ public void run()
         registerSourceRoots();
     }
 
+    private void preloadCommonPool()
+    {
+        try
+        {
+            // ensure common pool exists in the jvm
+            final Class<?> fjp = Thread.currentThread().getContextClassLoader()
+                    .loadClass("java.util.concurrent.ForkJoinPool");
+            final ExecutorService es = ExecutorService.class.cast(
+                    fjp
+                            .getMethod( "commonPool" )
+                            .invoke( null ) );
+            final int max = preloadCommonPool > 0
+                    ? preloadCommonPool :
+                    Integer.class.cast( fjp.getMethod( "getCommonPoolParallelism" ).invoke( null ) );
+            final CountDownLatch preLoad = new CountDownLatch( 1 );
+            for ( int i = 0;
+                  i < max;
+                  i++ )
+            {
+                es.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        try
+                        {
+                            preLoad.await();
+                        }
+                        catch ( InterruptedException e )
+                        {
+                            Thread.currentThread().interrupt();
+                        }
+                    }
+                });
+            }
+            preLoad.countDown();
+        }
+        catch (final Exception e)
+        {
+            getLog().debug(e.getMessage() + ", skipping commonpool earger init");
+        }
+    }
+
```

Comment:
After our chat, I think I understood what is going on here. I am not sure, however, why reflection would be needed.
Anyway. Most of us might not be that familiar with what is going on here. A few words on the 'why' would help a lot:
*why* using reflection, *why* initializing it (maybe in conjunction with the ThreadGroup).
From my understanding, no reflection would be needed because the created JFP would not be part of the thread group.

Code comments on the "why" would be great for our maintainers! :)

Reply:
Moved to java 8 since plugin does not use java 7 anymore which drops the reflection need.
Code by itself does not use thread groups but they are inherited from the calling thread so if commonpool is created lazily in the context of the main (isolatedclassloader) then it would use this threadgroup in some JVM which would let the JVM hang.
This is what https://github.com/mojohaus/exec-maven-plugin/pull/199/files#diff-dd2fe98649dca0e18d01909e1fc59b14c562cf349cb49a2cf19957147dca5551R76 means, happy to get a better wording if you find one, also not sure of the default we should use but at least we have a clean way to solve that issue now.